test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java

```diff
@@ -335,82 +169,88 @@ protected Settings restClientSettings() {
-                doSection.execute(restTestExecutionContext);
+    private RestClient buildClient(Settings settings) {
+        RestClient.Builder builder = RestClient.builder(clusterHosts.toArray(new HttpHost[0])).setMaxRetryTimeoutMillis(30000)
+                .setRequestConfigCallback(requestConfigBuilder -> requestConfigBuilder.setSocketTimeout(30000));
+        String keystorePath = settings.get(TRUSTSTORE_PATH);
+        if (keystorePath != null) {
+            final String keystorePass = settings.get(TRUSTSTORE_PASSWORD);
+            if (keystorePass == null) {
+                throw new IllegalStateException(TRUSTSTORE_PATH + " is provided but not " + TRUSTSTORE_PASSWORD);
+            }
+            Path path = PathUtils.get(keystorePath);
+            if (!Files.exists(path)) {
+                throw new IllegalStateException(TRUSTSTORE_PATH + " is set but points to a non-existing file");
+            }
+            try {
+                KeyStore keyStore = KeyStore.getInstance("jks");
+                try (InputStream is = Files.newInputStream(path)) {
+                    keyStore.load(is, keystorePass.toCharArray());
+                }
+                SSLContext sslcontext = SSLContexts.custom().loadTrustMaterial(keyStore, null).build();
+                SSLConnectionSocketFactory sslConnectionSocketFactory = new SSLConnectionSocketFactory(sslcontext);
+                builder.setHttpClientConfigCallback(new SSLSocketFactoryHttpConfigCallback(sslConnectionSocketFactory));
+            } catch (KeyStoreException|NoSuchAlgorithmException|KeyManagementException|CertificateException|IOException e) {
+                throw new RuntimeException(e);
             }
-            logger.debug("end setup test [{}]", testCandidate.getTestPath());
         }
 
-        restTestExecutionContext.clear();
-
         try {
-            for (ExecutableSection executableSection : testCandidate.getTestSection().getExecutableSections()) {
-                executableSection.execute(restTestExecutionContext);
-            }
-        } finally {
-            logger.debug("start teardown test [{}]", testCandidate.getTestPath());
-            for (DoSection doSection : testCandidate.getTeardownSection().getDoSections()) {
-                doSection.execute(restTestExecutionContext);
+            try (ThreadContext threadContext = new ThreadContext(settings)) {
+                Header[] defaultHeaders = new Header[threadContext.getHeaders().size()];
+                int i = 0;
+                for (Map.Entry<String, String> entry : threadContext.getHeaders().entrySet()) {
+                    defaultHeaders[i++] = new BasicHeader(entry.getKey(), entry.getValue());
+                }
+                builder.setDefaultHeaders(defaultHeaders);
             }
-            logger.debug("end teardown test [{}]", testCandidate.getTestPath());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```

Comment:
Yeah - I guess it doesn't matter because every user of it will already throw the IOException. I'll remove this. Originally I was thinking that it'd be nice not to have to deal with the checked exception in the test. But we already have it anyway.